src/internal/storage/server.go

```diff
@@ -220,23 +225,48 @@ func (w *writer) Write(data []byte) (int, error) {
 
 func (s *Server) ReadFilesetCDR(request *storage.ReadFilesetRequest, server storage.Fileset_ReadFilesetCDRServer) error {
 	ctx := server.Context()
-	// TODO: Use task chain here.
-	return s.readFileset(ctx, request, func(f fileset.File) error {
+	cache, err := lru.New[string, string](128)
+	if err != nil {
+		panic(err)
+	}
+	taskChain := taskchain.New(ctx, semaphore.NewWeighted(int64(taskParallelism)))
+	if err := s.readFileset(ctx, request, func(f fileset.File) error {
 		var refs []*cdr.Ref
-		for _, dr := range f.Index().File.DataRefs {
-			ref, err := s.Chunks.CDRFromDataRef(ctx, dr)
-			if err != nil {
+		numRefs := len(f.Index().File.DataRefs)
+		for i, dataRef := range f.Index().File.DataRefs {
+			i, dataRef := i, dataRef
+			if err := taskChain.CreateTask(func(ctx context.Context) (func() error, error) {
+				ref, err := s.Chunks.CDRFromDataRef(ctx, dataRef, cache)
+				if err != nil {
+					return nil, err
+				}
+				return func() error {
+					refs = append(refs, ref)
+					if i == numRefs-1 {
+						// TODO: Move to cdr package?
+						ref := chunk.CreateConcatRef(refs)
+						return server.Send(&storage.ReadFilesetCDRResponse{
+							Path: f.Index().Path,
+							Ref:  ref,
+						})
+					}
+					return nil
+				}, nil
+			}); err != nil {
 				return err
 			}
-			refs = append(refs, ref)
 		}
-		// TODO: Move to cdr package?
-		ref := chunk.CreateConcatRef(refs)
-		return server.Send(&storage.ReadFilesetCDRResponse{
-			Path: f.Index().Path,
-			Ref:  ref,
-		})
-	})
+		if numRefs == 0 {
```

Comment:
This should be the first check in the callback. Also, I don't think there is really much value in having the `numRefs` variable, just getting the length in the two places it is referenced seems simpler.